Add a new protobuf codec for info data

LorenVS opened the pull request:
Adds support for a new protobuf codec. This will make it easier to write tools for analyzing dart2js output in other languages or environments.

I have currently only tested this code by converting an info.json to an Info, and then to an InfoPB. I would like to hook into dart2js to generate the InfoPB directly, but I'm sending this review out first to see if this is something we would consider accepting.

I'm including the proto file in the root of this repository. I'm also unaware (externally) how to generate the dart protoc output automatically on build. Given that this would take a dependency on protoc, it might not be trivial? I'm currently just submitting the generated proto files with the rest of this change, not sure whether that is acceptable.

The JSON codec currently encodes compilationMoment as a string. I could do something similar with this codec, which would break the dependency on fixnum, since it would likely be safe to encode the durations as int32. Please let me know if that would be preferable.

Commit:
Add a new protobuf codec for dart2js info.

Currently only supports conversion from Info to protobuf, and not in
reverse. This is intended to make writing analysis tools in other
languages significantly easier.

Commit:
Support null values on a variety of Input fields.

Commit:
Add experimental comment on AllInfoProtoCodec

harryterkelsen commented on info.proto:
```diff
@@ -0,0 +1,265 @@
+  /** The dependency element's serialized_id, references as FunctionInfo or FieldInfo. */
+  string target_id = 1;
+
+  /** Either a selector mask indicating how ths is used, or 'inlined'. */
```
ths -> this

harryterkelsen commented on info.proto:
```diff
@@ -0,0 +1,265 @@
+  /** Summary information about the program. */
+  ProgramInfoPB program = 1;
+
+  /** All the recorded infromation about elements processed by the compiler. */
```
infromation -> information

harryterkelsen commented on info.proto:
```diff
@@ -0,0 +1,265 @@
+  /** An id to uniquely identify this info among infos of the same kind. */
+  int32 id = 2;
+
+  /** A globally unique id which combining kind and id together. */
```
combining -> combines

Commit:
Run dartfmt

harryterkelsen commented on lib/proto_info_codec.dart:
```diff
@@ -0,0 +1,258 @@
+    // analysis in non-Dart languages. In these cases, its likely that the
+    // proto format itself will be more useful to consumers than trying to
+    // convert back to the [AllInfo] graph.
+    throw new UnimplementedError('ProtoToAllInfoConverter is not implemented');
```
proto -> AllInfo is a good test that all of the required information is contained in the proto. I would recommend implementing it

LorenVS replied:
I'm still missing some information from AllInfo at the moment. Specifically the holdings and dependencies are confusing me a bit. Also deferredFiles is currently just a Map, and I need to figure out how to model it as a proto. Is it alright if I look at this implementation for round tripping after I've figured out some of this other stuff?

harryterkelsen replied:
That's fine, please add a TODO

harryterkelsen commented on pubspec.yaml:
```diff
@@ -12,7 +12,9 @@ dependencies:
   args: '>=0.13.0 <2.0.0'
   charcode: ^1.1.0
   collection: ^1.10.1
+  fixnum: '>=0.10.7'
```
the version of fixnum in the Dart SDK is 10.5. You should set it to be 10.5 here too, unless you need 10.7 features, in which case you will need to also roll the version of fixnum in the Dart SDK.

sigmundch replied:
also note: we should use `^0.10.5` so that we don't accidentally depend on a newer version with breaking changes.

harryterkelsen reviewed (commented):
Looks mostly good. It could be possible to make dart2js generate this version of dump-info directly, probably behind a flag like `--dump-info-format` but it might require pulling in proto dependencies if they are needed to serialize proto to disk.

hterkelsen commented:
@sigmundch FYI

Commit:
Change proto_info_codec to be its own import.

This should prevent dart2js from importing the generated protos, the
protobuf library and the fixnum library by default, until we have a
chance to verify this change's impact on the snapshot size.

Commit:
Address initial review comments.

LorenVS commented:
I am hoping to get this into dart2js as a proto output format. Sigmund wants to run some tests to verify that this doesn't negatively impact the snapshot size first. In order to make this change without that risk, he recommended I move the new proto codec to a separate import, which I've done.

Commit:
Add a utility binary for converting info.json files to info.pb

Commit:
Change the comment on ProtoToAllInfoConverter to a TODO.

LorenVS commented:
I've changed the comment on ProtoToAllInfoConverter to a TODO, and also added an executable for converting info.json files to info.pb files. PTAL

hterkelsen commented:
LGTM, okay to merge once Siggi has determined that this doesn't negatively impact snapshot size

sigmundch commented on bin/info_json_to_proto.dart:
```diff
@@ -0,0 +1,24 @@
+
+main(args) async {
+  if (args.length != 2) {
+    print('usage: dart tool/library_size_split.dart '
```
update the usage message (info_json_to_proto.dart)

sigmundch reviewed (approved):
Thanks for making all the changes Loren.

Now that the import is separate, and dart2js is not importing this by default, we should be OK in terms of the snapshot size. 

I just filed https://github.com/dart-lang/dart2js_info/issues/36 to track more details about the offline discussions I had with @LorenVS

Commit:
Update usage comment for info_json_to_proto.
